qualtran/bloqs/arithmetic/sorting.py

```diff
@@ -48,82 +64,331 @@ class Comparator(Bloq):
+
+    This is used by `BitonicMerge` to apply parallel merges with offsets 1, 2, 4 and so on.
+
+    Args:
+        k: size of the input list.
+        offset: compare numbers whose indices are offset apart.
+        bitsize: value of $n$ (i.e. the inputs are $n$-bit numbers).
+
+    Registers:
+        xs: input list of numbers.
+        junk (RIGHT): ancilla generated by comparators.
+    """
+
+    k: SymbolicInt
+    offset: SymbolicInt
+    bitsize: SymbolicInt
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature(
+            [
+                Register("xs", QUInt(self.bitsize), shape=(self.k,)),
+                Register("junk", QBit(), shape=(self.num_comparisons,), side=Side.RIGHT),
+            ]
+        )
+
+    def is_symbolic(self):
+        return is_symbolic(self.bitsize, self.k, self.offset)
+
+    @cached_property
+    def num_comparisons(self) -> SymbolicInt:
+        """Number of `Comparator` gates used in the decomposition"""
+        if is_symbolic(self.k, self.offset):
+            return self.k // 2  # upper bound
+
+        full = (self.k // (2 * self.offset)) * self.offset
+        rest = self.k % (2 * self.offset)
+        return full + max(rest - self.offset, 0)
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', xs: 'SoquetT') -> Dict[str, 'SoquetT']:
+        if is_symbolic(self.k, self.offset):
+            raise DecomposeTypeError(f"Cannot decompose symbolic {self=}")
+
+        # make mypy happy
+        k = int(self.k)
+        offset = int(self.offset)
+        assert isinstance(xs, np.ndarray)
+
+        comp = Comparator(self.bitsize)
+
+        junk = []
+        for start in range(0, k, 2 * offset):
+            for step in range(offset):
+                i, j = start + step, start + step + offset
+                if j >= k:
+                    break
+                xs[i], xs[j], anc = bb.add(comp, a=xs[i], b=xs[j])
+                junk.append(anc)
+
+        assert len(junk) == self.num_comparisons
+
+        return {'xs': xs, 'junk': np.array(junk)}
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        return {(Comparator(self.bitsize), self.num_comparisons)}
+
+
+@bloq_example
+def _parallel_compare() -> ParallelComparators:
+    parallel_compare = ParallelComparators(7, 2, bitsize=3)
+    return parallel_compare
+
+
+_PARALLEL_COMPARATORS_DOC = BloqDocSpec(
+    bloq_cls=ParallelComparators,
+    examples=[_parallel_compare],
+    import_line="from qualtran.bloqs.arithmetic.sorting import ParallelComparators",
+)
+
+
+@frozen
+class BitonicMerge(Bloq):
+    r"""Merge two sorted sequences of n-bit integers.
+
+    Given two sorted lists of length `half_length`, merges them inplace into a single
+    sorted list.
+
+    Currently only supports `half_length` equal to a power of two (#1090).
+
+    If each half has length $k$, then the merge network uses $k * (\log{k} + 1)$ comparisons
+    when $k$ is a power of 2.
```

Comment:
Please also add information / $$..$$ formula to specify what the `junk` stores since it can be potentially useful depending upon the context.

Reply:
Added a brief description (similar to the one in `BitonicSort`).

Did you also want a description of the order of the ancillas?

Reply:
The order is not as important, but we should at least specify that the ancilla qubits store the results of intermediate comparisons.